### ISSUE TITLE
Occasional lags/ BSOD/Disk Corruption

### ISSUE THREAD

**Issue by lellis1936** [r0m30/msed#44]
I seem to have experienced a few weird behaviors in Windows 7 Ultimate after issuing the following two commands:

msed --setMBREnable on <password> \\.\physicaldrive0
msed --enableLockingRange 0 <password> \\.\physicaldrive0

I am not blaming your software but felt it my duty to report it.

I frequently run these two commands because I often swap between locked and unlocked mode.

I always do a cold shutdown immediately after running them, but on some occasions the OS has hung shutting down, BSOD, or reported "chkdsk needed" on restart and I have seen some NTFS corruption as well.

Should these two commands always be completely safe while Windows is running?  I typically run Samsung's RAPID mode which may or may not complicate matters, though I have *never* seen it cause a problem.

Finally, my system is generally solid as a rock in every respect which is why I thought I should post this.  My SSD has never reported any errors in SMART (Samsung EVO 250 Pro).

At this point I think I will avoid toggling much between locked and unlocked mode.  I am wondering whether the switch might be triggering some odd/buggy behavior in the drive's firmware.

Thank you for this great software, it's really working well for me in all other respects.

**Comment by r0m30** [comment]
Hi, 
The first thing I'd do is issue 
setMBRDone on 
and 
setLockingRange 0 RW 
before the two commands you are using.

These two commands will insure that the drive is read/write and not shadowed when you issue the two commands you are currently using.

If at some point you have power cycled the drive that would reset MBRDone to off.  If MBRDone is off and MBREnable is on the drive will present the shadow MBR not the real disk LBAs.  Then you tell Windows to shut down and it tries to flush it's disk cache, unfortunately that would be written to the shadow MBR and discarded.   I believe Samsung's rapid mode is yet another level of caching that would also not be properly flushed if the disk was in shadow MBR mode. 

I'm not sure what the status of the read/write locking flags are when you power cycle with a locking range disabled but it could be that they are locked causing the same behavior described above except that the drive should return errors when is is locked, so issuing the second command will eliminate any doubt about the locking state of the drive before you enable locking.

**Comment by lellis1936** [comment]
I've confirmed through testing that the scenario you described is the culprit.

I have been running two sets of commands to toggle between locked and unlocked mode:

To unlock:
msed --disableLockingRange 0 xx \\.\physicaldrive0
msed --setMBREnable off xx \\.\physicaldrive0

To lock:
msed --setMBREnable on xx  \\.\physicaldrive0
msed --enableLockingRange 0 xx \\.\physicaldrive0

The problem is the second set:  typically, MBRDone would *always* be off when those commands are issued, leading to the shutdown problems described.  The only time this would not be the case is after the initialize sequence you recommend, since that appears to set MDRDone on.

So I have added this to the lock sequence, and it appears to solve the problem:
msed --setMBRDone on xx \\.\physicaldrive0

But, I am wondering whether MBRDone could/should be forced on as a side-effect of one or both of the two lock commands.  This might avoid the risk of BSODs/disk corruption.

**Comment by r0m30** [comment]
It could.....
I think of the basic state manipulation commands as primitives, and then add convenience commands like intitialsetup for "standard" use cases.  I dislike it when a program assumes it knows what I want to do.
Still not sure that it should....

**Comment by lellis1936** [comment]
I understand exactly what you're saying.

Your documented procedure for removing locking is:

msed --disableLockingRange 0 xx \\.\physicaldrive0
msed --setMBREnable off xx \\.\physicaldrive0

You might consider just adding the followup sequence to the documentation:
msed --setMBRDone on xx \\.\physicaldrive0
msed --setMBREnable on xx \\.\physicaldrive0
msed --enableLockingRange 0 xx \\.\physicaldrive0

If alternatively a user completely removes Opal through and then adds it back through the initial setup process, you describe, he will have MBRDone on as well.   

I guess I just can't see a case where you would enable MBR and not also want MBRDone in a running Windows system (or probably a Linux system either) since it seems that it can only result in an unstable system until a reboot occurs.  And even an immediate shutdown attempt can then fail, as has happened to me a number of times.  So you might consider refusing to setMBREnable on unless MDRDone is also on (issuing an appropriate error message if an attempt is made).

Please close this issue out if you like; thanks for hearing me out.

**Comment by brendanhoar** [comment]
On Wed, Oct 21, 2015 at 1:59 PM, lellis1936 <notifications@github.com>
wrote:
>
> I guess I just can't see a case where you would enable MBR and not also
> want MBRDone in a running Windows system (or probably a Linux system
> either) since it seems that it can only result in an unstable system until
> a reboot occurs. And even an immediate shutdown attempt can then fail, as
> has happened to me a number of times. So you might consider refusing to
> setMBREnable on unless MDRDone is also on (issuing an appropriate error
> message if an attempt is made).
>

This instability issue is true for system/boot drives using the primitives
originally specified.

However, this may not necessarily be true of secondary drives, where one
would (should?) dismount the active volumes first (and therefore flush the
cache and mark the volume closed) before performing OPAL operations. If a
contemporary OS is on the ball, it should be able to handle any locked
state changes, read vs. read/write capabilities, or altered geometries that
occur after OPAL commands on drives that contain no mounted volumes.

That would lead me (just a user) to suggest that the updating the docs to
issue the existing primitives in the correct order is probably the best
approach.

Brendan

**Comment by r0m30** [comment]
@lellis1936   I've updated the doc on the Drive-Trust-Alliance/sedutil wiki (that's where the code and doc will be going forward).  I'm not ready to close this because I have that feeling that this is an issue that may require that I err on the side of safety over design philosophy.

@brendanhoar   I have updated the doc as I said above, the issue I'm having is that most people, myself included, usually only skim the doc if they read it at all.  This could lead to data loss and/or corruption and people will blame the SW not themselves no matter how well the danger is documented..
